### PR TITLE
Assembly line data stick fix

### PR DESCRIPTION
**Pull request by draknyte1**
Should fix Data Stick recipe mismatches.

### CODE REVIEW

**Comment by draknyte1** [comment]
I might clean up recipe handling for the scanner and assembly line.

File: src/main/java/gregtech/api/util/GT_AssemblyLineUtils.java [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+	}
+
+	/**
+	 * @param aRecipe - The recipe to generate a Recipe Hash String from.
+	 * @return The Recipe Hash String.
+	 */
+	public static String generateRecipeHash(GT_Recipe_AssemblyLine aRecipe) {
+		String aHash = "Invalid.Recipe.Hash";		
+		if (aRecipe != null) {
+			aHash = "Hash."+aRecipe.hashCode();
```
Comment (Glease) [inline src/main/java/gregtech/api/util/GT_AssemblyLineUtils.java:213]: `GT_Recipe_AssemblyLine` or `GT_Recipe` does not override `hashCode()`. Your hash will change across game restarts, basically renders this hash useless.

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java [hunk 2]
```diff
@@ -156,117 +158,111 @@ public boolean isCorrectMachinePart(ItemStack aStack) {
...
-                    tStack[i] = GT_Utility.loadItem(tTag, "" + i);
-                    if (tStack[i] == null) {
-                        flag = false;
-                        continue;
-                    }
-                    if (GT_Values.D1)
-                        GT_FML_LOGGER.info("Item " + i + " : " + tStack[i].getUnlocalizedName());
-                    if (GT_Utility.areStacksEqual(tStack[i], stackInSlot, true) && tStack[i].stackSize <= stackInSlot.stackSize) {
-                        flag = false;
+            	if (i >= aItemCount) {
```
Comment (Glease) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java:195]: Somehow make it not use a mix of whitespace or tab please. Either is ok to me, just please don't mix it.

**Review by Glease — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java (hunk 2 above)
Comment (draknyte1) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java:195]: I can just reformat the entire class? That'll fix it any make it harder diff (Why I avoided it and only formatted the block I changed)

Comment (Glease) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java:195]: Weird... Let it be then. I'm not sure what's wrong with eclipse's reformatter. We can fix it next time someone need to touch it.

File: src/main/java/gregtech/api/util/GT_Recipe.java [hunk 3]
```diff
@@ -564,6 +564,77 @@ public GT_Recipe_AssemblyLine(ItemStack aResearchItem, int aResearchTime, ItemSt
         	mEUt = aEUt;
         	mOreDictAlt = aAlt;
         }
+
+		@Override
+		public int hashCode() {
+			final int prime = 31;
+			int result = 1;			
+			GT_ItemStack[] thisInputs = new GT_ItemStack[this.mInputs.length];
+			int inputHash = Arrays.deepHashCode(thisInputs);
```
Comment (Glease) [inline src/main/java/gregtech/api/util/GT_Recipe.java:573]: Uh, why are you hashing an empty array you just created one instructions ago?

**Review by Glease — CHANGES_REQUESTED** [review]

**Comment by D-Cysteine** [comment]
I noticed that you're using `GT_ItemStack` for hashing in a few places:

 * https://github.com/GTNewHorizons/GT5-Unofficial/blob/25d664c62e4b658f1bc3c41706488ab351fec2db/src/main/java/gregtech/api/util/GT_AssemblyLineUtils.java#L24
 * https://github.com/GTNewHorizons/GT5-Unofficial/blob/25d664c62e4b658f1bc3c41706488ab351fec2db/src/main/java/gregtech/api/util/GT_Recipe.java#L572-L573

I believe that `GT_ItemStack` ignores both NBT and stack size, so we may get unexpected behavior if we ever have assembly line recipes which differ only in NBT or stack size. Do we need to handle that case? If so, I guess we'd have to either modify `GT_ItemStack` to support that (which may break a bunch of stuff as it's used in a lot of places I think), or create a new wrapper class. I did this to get hash support with minimal effort, though it doesn't handle stack size:
https://github.com/GTNewHorizons/GT5-Unofficial/blob/1dd66efceca26d9931010adb24932bba408ca630/src/main/java/gregtech/api/util/GT_Single_Recipe_Check.java#L260

Wanted to mention this because I was concerned that, even if this works fine currently, if we ever do introduce recipes in the future which run into this case, it might be very difficult to debug misbehaving hash maps.

**Comment by draknyte1** [comment]
> I noticed that you're using `GT_ItemStack` for hashing in a few places:
> 
> * https://github.com/GTNewHorizons/GT5-Unofficial/blob/25d664c62e4b658f1bc3c41706488ab351fec2db/src/main/java/gregtech/api/util/GT_AssemblyLineUtils.java#L24
> * https://github.com/GTNewHorizons/GT5-Unofficial/blob/25d664c62e4b658f1bc3c41706488ab351fec2db/src/main/java/gregtech/api/util/GT_Recipe.java#L572-L573
> 
> I believe that `GT_ItemStack` ignores both NBT and stack size, so we may get unexpected behavior if we ever have assembly line recipes which differ only in NBT or stack size. Do we need to handle that case? If so, I guess we'd have to either modify `GT_ItemStack` to support that (which may break a bunch of stuff as it's used in a lot of places I think), or create a new wrapper class. I did this to get hash support with minimal effort, though it doesn't handle stack size:
> 
> https://github.com/GTNewHorizons/GT5-Unofficial/blob/1dd66efceca26d9931010adb24932bba408ca630/src/main/java/gregtech/api/util/GT_Single_Recipe_Check.java#L260
> 
> Wanted to mention this because I was concerned that, even if this works fine currently, if we ever do introduce recipes in the future which run into this case, it might be very difficult to debug misbehaving hash maps.

It does, but if the only thing differing is stack size or NBT, you’d have a recipe conflict anyway. So it doesn’t really matter. GT_ItemStack is suitable for hashing when comparing recipes or getting a hash for an AL recipe imho.

**Comment by D-Cysteine** [comment]
Okay, sounds good. Yeah, I'm not too familiar with what constraints we have on assembly line recipes, so I wasn't sure if this was actually an issue or not.

The specific cases I had in mind were something like:
 * If two recipes differ only in a single input ingredient, which is something like a universal cell or an enchanted book that stores its attributes in NBT
 * If two recipes have the same output, just with different stack sizes, like if we wanted to add a x64 version of a smaller recipe or something

which I think we actually do have examples of currently, but in other recipe maps.

**Comment by draknyte1** [comment]
> which I think we actually do have examples of currently, but in other recipe maps.

I believe the inputs differing so slightly would lead to recipe conflicts. 

To my knowledge AL outputs only have a recipe. This PR wholly assumes as much.

**Comment by D-Cysteine** [comment]
I do know that the regular recipe input check handles NBT:
 * https://github.com/GTNewHorizons/GT5-Unofficial/blob/experimental/src/main/java/gregtech/api/util/GT_Recipe.java#L419

(the `false` there is the value for `aIgnoreNBT`)

so we do need to check it in general, but might not need it for the specific case of assembly line

**Comment by D-Cysteine** [comment]
Although, come to think of it, the recipe hash is how we detect that a recipe has been changed in the pack, right? So any changes to the recipe must necessarily result in a change to the hash. Then we probably need to check stack size, since a recipe might have only the stack sizes of its inputs changed?

**Review by Glease — APPROVED** [review]

**Pull request merged**